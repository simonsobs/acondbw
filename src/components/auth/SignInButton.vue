<template>
  <span>
    <span v-if="token && user">
      <v-menu bottom offset-y v-model="menu">
        <template v-slot:activator="{ on, attr }">
          <v-btn icon class="no-uppercase" v-bind="attr" v-on="on">
            <v-avatar size="36">
              <img :src="user.avatarUrl" :alt="user.login" />
            </v-avatar>
          </v-btn>
        </template>
        <v-list>
          <v-list-item>
            <v-list-item-content>
              <v-list-item-title>{{ user.name }}</v-list-item-title>
              <v-list-item-subtitle>{{ user.login }}</v-list-item-subtitle>
            </v-list-item-content>
            <v-list-item-icon>
              <v-icon left>mdi-github</v-icon>
            </v-list-item-icon>
          </v-list-item>
          <v-divider></v-divider>
          <v-dialog v-model="dialog" max-width="600">
            <template v-slot:activator="{ on }">
              <v-list-item v-on="on">
                <v-list-item-icon>
                  <v-icon>mdi-logout</v-icon>
                </v-list-item-icon>
                <v-list-item-content>
                  <v-list-item-title>Sign Out</v-list-item-title>
                </v-list-item-content>
              </v-list-item>
            </template>
            <sign-out-confirmation v-on:finished="dialog = false; menu = false"></sign-out-confirmation>
          </v-dialog>
        </v-list>
      </v-menu>
    </span>
    <v-btn v-if="!token" depressed :to="pathToSignIn">
      <v-icon left>mdi-github</v-icon>
      Sign In
    </v-btn>
  </span>
</template>

<script>
import GitHubUser from "@/graphql/auth/GitHubUser.gql";
import SignOutConfirmation from "./SignOutConfirmation";

export default {
  name: "SignInButton",
  components: { SignOutConfirmation },
  data: () => ({
    githubUser: null,
    menu: false,
    dialog: false
  }),
  apollo: {
    githubUser: {
      query: GitHubUser,
      skip: function() {
        return !this.token;
      },
      result(result) {
        if (!result.data || !result.data.githubUser) {
          this.$store.dispatch("unsetToken");
          return;
        }
      }
    }
  },
  computed: {
    token() {
      return this.$store.state.token;
    },
    user() {
      if (!this.token) {
        return null;
      } else if (!this.githubUser) {
        return null;
      } else {
        return this.githubUser;
      }
    },
    pathToSignIn() {
      return this.$router.resolve({ name: "SignIn" }).route.path;
      // i.e., "/signin"
    }
  },
  watch: {
    token: function() {
      if (!this.token) return;
      this.$apollo.queries.githubUser.refetch();
    }
  }
};
</script>

<style scoped>
.no-uppercase {
  text-transform: none;
}
</style>