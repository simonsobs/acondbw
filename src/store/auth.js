import { onLogin, onLogout, AUTH_TOKEN } from "@/vue-apollo";
import GitHubUser from "@/graphql/auth/GitHubUser.gql";

function createInitialState() {
  const token = JSON.parse(localStorage.getItem(AUTH_TOKEN));
  const githubUser = JSON.parse(localStorage.getItem("github-user"));
  const ret = {
    token: token,
    githubUser: githubUser
  }
  return ret;
}

const initialState = createInitialState();

export const auth = {
  state: initialState,
  mutations: {
    set_token(state, token) {
      state.token = token;
    },
    set_github_user(state, githubUser) {
      state.githubUser = githubUser;
    }
  },
  actions: {
    async setToken({ commit }, { token, apolloClient }) {
      await onLogin(apolloClient, token);
      commit("set_token", token);
    },
    async signOut({ commit }, apolloClient) {
      localStorage.removeItem("github-user");
      commit("set_github_user", null);
      await onLogout(apolloClient);
      commit("set_token", null);
    },
    async loadGitHubUser({ commit, dispatch }, { apolloClient }) {
      try {
        const { data } = await apolloClient.query({ query: GitHubUser });
        const githubUser = data.githubUser;
        localStorage.setItem("github-user", JSON.stringify(githubUser));
        commit("set_github_user", githubUser);
      } catch (error) {
        dispatch("signOut", apolloClient);
        throw error;
      }
    },
  },
};
