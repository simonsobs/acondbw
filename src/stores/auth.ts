import { defineStore } from "pinia";

import {
  signIn,
  signOut,
  restoreFromLocalStorage,
  isSignedIn,
} from "@/utils/auth";

interface GitHubViewer {
  login: string;
  name?: string;
  avatar_url?: string;
}

interface SignInInfo {
  gitHubViewer: GitHubViewer | null;
  isSignedIn: boolean;
  isAdmin: boolean;
}

// https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/
interface OAuthError {
  error: string;
  error_description: string;
  error_uri: string;
}

interface AuthStoreState extends SignInInfo {
  token: string | null;
  lastError: OAuthError | null;
}

export const useAuthStore = defineStore("auth", {
  state: (): AuthStoreState => {
    const initialState: AuthStoreState = {
      token: null,
      gitHubViewer: null,
      isSignedIn: false,
      isAdmin: false,
      lastError: null,
    };
    try {
      const { token, signInInfo } = restoreFromLocalStorage();
      return {
        ...initialState,
        token,
        ...signInInfo,
      };
    } catch (error) {
      return initialState;
    }
  },
  actions: {
    setSignInInfo(info) {
      this.gitHubViewer = info.gitHubViewer;
      this.isSignedIn = info.isSignedIn;
      this.isAdmin = info.isAdmin;
    },
    resetSignInInfo() {
      this.gitHubViewer = null;
      this.isSignedIn = false;
      this.isAdmin = false;
    },
    async checkIfSignedIn(apolloClient) {
      try {
        const result = await isSignedIn(apolloClient);
        if (result) {
          this.setSignInInfo(result);
          return;
        }
      } catch (error) {}
      this.signOut(apolloClient);
    },
    async signOut(apolloClient) {
      this.resetSignInInfo();
      this.token = null;
      await signOut(apolloClient);
    },
    async signIn({ code, state, apolloClient }) {
      this.lastError = null;
      try {
        const { token, signInInfo } = await signIn(code, state, apolloClient);
        this.token = token;
        this.setSignInInfo(signInInfo);
      } catch (error) {
        this.signOut(apolloClient);
        throw error;
      }
    },
    setRequestAuthError(query) {
      const error: OAuthError = (({ error, error_description, error_uri }) => ({
        error,
        error_description,
        error_uri,
      }))(query);
      this.lastError = error;
    },
    clearAuthError() {
      this.lastError = null;
    },
  },
});
