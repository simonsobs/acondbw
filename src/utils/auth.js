const querystring = require("querystring");
const cryptoRandomString = require("crypto-random-string");
import _ from "lodash";

import { apolloClient, onLogin, onLogout, AUTH_TOKEN } from "@/vue-apollo";
export const AUTH_STATE = "auth-state";

import GitHubOAuthAppInfo from "@/graphql/queries/GitHubOAuthAppInfo.gql";
import AuthenticateWithGitHub from "@/graphql/mutations/AuthenticateWithGitHub.gql";
import GitHubViewer from "@/graphql/queries/GitHubViewer.gql";
import SignInInfo from "@/graphql/queries/SignInInfo.gql";
import IsSignedIn from "@/graphql/queries/IsSignedIn.gql";

/**
 *
 * @param {object} window - the window object
 * @param {object} apolloClient - an apollo client
 * @param {object} callbackRoute - a route to be redirected from the rigistered callback route
 * @param {string} scope - scopes as described in https://docs.github.com/en/developers/apps/scopes-for-oauth-apps
 */
export async function redirectToGitHubAuthURL(
  window,
  apolloClient,
  callbackRoute,
  scope
) {
  try {
    const { data } = await apolloClient.query({ query: GitHubOAuthAppInfo });
    const gitHubOAuthAppInfo = data.gitHubOAuthAppInfo;
    const code = cryptoRandomString({ length: 8, type: "url-safe" });
    const state = btoa(
      JSON.stringify({
        redirect: callbackRoute,
        code: code,
      })
    );
    const params = {
      response_type: "code",
      client_id: gitHubOAuthAppInfo.clientId,
      redirect_uri: gitHubOAuthAppInfo.redirectUri,
      scope: scope,
      state: state,
    };
    let queryString = querystring.stringify(params);
    const uri = gitHubOAuthAppInfo.authorizeUrl + "?" + queryString;
    localStorage.setItem(AUTH_STATE, JSON.stringify(state));
    window.location.href = uri;
  } catch (error) {
    localStorage.removeItem(AUTH_STATE);
    throw error;
  }
}

/**
 * Validate the state by comparing the one stored in localStorage
 * @param {string} state - the state returned from the auth server
 * @returns {boolean} true if validated
 */
export function validateState(state) {
  if (!state) {
    return false;
  }
  const authState = JSON.parse(localStorage.getItem(AUTH_STATE));
  if (!authState) {
    return false;
  }
  if (!(authState == state)) {
    return false;
  }
  return true;
}

export function restoreFromLocalStorage() {
  try {
    const token = JSON.parse(localStorage.getItem(AUTH_TOKEN));
    const signInInfo = JSON.parse(localStorage.getItem("sign-in-info"));

    if (token && signInInfo) {
      return { token, signInInfo };
    }
  } catch {
    localStorage.clear();
  }

  localStorage.removeItem(AUTH_TOKEN);
  localStorage.removeItem("sign-in-info");
  // do not call localStorage.clear() here because it will delete
  // AUTH_STATE when redirected from the auth server

  return {
    token: null,
    signInInfo: { gitHubViewer: null, isSignedIn: false, isAdmin: false },
  };
}

export async function isSignedIn(apolloClient) {
  try {
    const { data } = await apolloClient.query({ query: IsSignedIn });
    if (data.isSignedIn) {
      const signInInfo = await getSignInInfo();
      localStorage.setItem("sign-in-info", JSON.stringify(signInInfo));
      return signInInfo;
    }
  } catch {}
  await signOut(apolloClient);
  return false;
}

export async function signOut(apolloClient) {
  await onLogout(apolloClient);
  localStorage.clear();
}

export async function signIn(code, state, apolloClient) {
  try {
    const token = await exchangeCodeForToken(code, state, apolloClient);
    await onLogin(apolloClient, token);
    const signInInfo = await getSignInInfo();
    localStorage.setItem("sign-in-info", JSON.stringify(signInInfo));
    return { token, signInInfo };
  } catch (error) {
    await signOut(apolloClient);
    throw error;
  }
}

export async function exchangeCodeForToken(code, state, apolloClient) {
  if (!validateState(state)) {
    throw new Error("The state was invalid.");
  }
  const { data } = await apolloClient.mutate({
    mutation: AuthenticateWithGitHub,
    variables: { code: code },
  });
  const authPayload = data.authenticateWithGitHub.authPayload;
  const token = JSON.stringify(authPayload.token);
  return token;
}

export async function getSignInInfo() {
  const { data } = await apolloClient.query({ query: SignInInfo });
  return _.pick(data, ["isSignedIn", "isAdmin", "gitHubViewer"]);
}
