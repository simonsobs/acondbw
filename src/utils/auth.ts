import VueRouter, { Route, RawLocation, Location } from "vue-router";
import { DollarApollo } from "vue-apollo/types/vue-apollo";
import { v4 as uuidv4 } from "uuid";
import _ from "lodash";

import { apolloClient, onLogin, onLogout, AUTH_TOKEN } from "@/vue-apollo";
export const AUTH_STATE = "auth-state";

import QUERY_GIT_HUB_O_AUTH_APP_INFO from "@/graphql/queries/GitHubOAuthAppInfo.gql";
import MUTATE_AUTHENTICATE_WITH_GIT_HUB from "@/graphql/mutations/AuthenticateWithGitHub.gql";
import QUERY_SIGN_IN_INFO from "@/graphql/queries/SignInInfo.gql";
import QUERY_IS_SIGNED_IN from "@/graphql/queries/IsSignedIn.gql";

export interface GitHubViewer {
  login: string;
  name?: string;
  avatar_url?: string;
}

export interface SignInInfo {
  gitHubViewer: GitHubViewer | null;
  isSignedIn: boolean;
  isAdmin: boolean;
}

// OAuth2 authorization request parameters
// https://www.oauth.com/oauth2-servers/authorization/the-authorization-request/
type RequestParams = {
  response_type: "code";
  client_id: string;
  redirect_uri?: string;
  scope?: string;
  state?: string;
};

export interface UnencodedState {
  redirect: Location;
  randomString: string;
}

/**
 *
 * @param apolloClient - an apollo client
 * @param callbackRoute - a route to be redirected from the registered callback route
 * @param scope - scopes as described in https://docs.github.com/en/developers/apps/scopes-for-oauth-apps
 */
export async function redirectToGitHubAuthURL(
  apolloClient: DollarApollo<any>,
  callbackRoute: Location,
  scope: string
) {
  const rawState: UnencodedState = {
    redirect: callbackRoute,
    randomString: uuidv4(),
  };
  const state = encodeState(rawState);
  try {
    const { data } = await apolloClient.query({
      query: QUERY_GIT_HUB_O_AUTH_APP_INFO,
    });
    const gitHubOAuthAppInfo = data.gitHubOAuthAppInfo;
    const params: RequestParams = {
      response_type: "code",
      client_id: gitHubOAuthAppInfo.clientId,
      redirect_uri: gitHubOAuthAppInfo.redirectUri,
      scope: scope,
      state: state,
    };
    const queryString = new URLSearchParams(params).toString();
    const uri = gitHubOAuthAppInfo.authorizeUrl + "?" + queryString;
    localStorage.setItem(AUTH_STATE, JSON.stringify(state));
    window.location.assign(uri);
  } catch (error) {
    localStorage.removeItem(AUTH_STATE);
    throw error;
  }
}

function encodeState(rawState: UnencodedState) {
  const jsonString = JSON.stringify(rawState);
  return btoa(jsonString);
}

export function validateAndDecodeState(state: any) {
  if (typeof state !== "string") return null;
  if (!validateState(state)) return null;
  return decodeState(state);
}

function decodeState(state: string) {
  const jsonString = atob(state);
  const rawState: UnencodedState = JSON.parse(jsonString);
  return rawState;
}

/**
 * Validate the state by comparing the one stored in localStorage
 * @param state - the state returned from the auth server
 * @returns true if validated
 */
export function validateState(state: string) {
  if (!state) return false;
  const stored = localStorage.getItem(AUTH_STATE);
  if (!(stored === JSON.stringify(state))) return false;
  return true;
}

export function restoreFromLocalStorage() {
  try {
    const tokenJson = localStorage.getItem(AUTH_TOKEN);
    const signInInfoJson = localStorage.getItem("sign-in-info");
    const token = tokenJson && (JSON.parse(tokenJson) as string);
    const signInInfo =
      signInInfoJson && (JSON.parse(signInInfoJson) as SignInInfo);

    if (token && signInInfo) {
      return { token, signInInfo };
    }
  } catch {
    localStorage.clear();
  }

  localStorage.removeItem(AUTH_TOKEN);
  localStorage.removeItem("sign-in-info");
  // do not call localStorage.clear() here because it will delete
  // AUTH_STATE when redirected from the auth server

  return {
    token: null as string | null,
    signInInfo: {
      gitHubViewer: null,
      isSignedIn: false,
      isAdmin: false,
    } as SignInInfo,
  };
}

export async function isSignedIn(apolloClient: DollarApollo<any>) {
  try {
    const { data } = await apolloClient.query({ query: QUERY_IS_SIGNED_IN });
    if (data.isSignedIn) {
      const signInInfo = await getSignInInfo();
      localStorage.setItem("sign-in-info", JSON.stringify(signInInfo));
      return signInInfo;
    }
  } catch {}
  await signOut(apolloClient);
  return false;
}

export async function signOut(apolloClient: DollarApollo<any>) {
  await onLogout(apolloClient);
  localStorage.clear();
}

export async function signIn(
  code: string,
  state: string,
  apolloClient: DollarApollo<any>
) {
  try {
    const token = await exchangeCodeForToken(code, state, apolloClient);
    await onLogin(apolloClient, token);
    const signInInfo = await getSignInInfo();
    localStorage.setItem("sign-in-info", JSON.stringify(signInInfo));
    return { token, signInInfo };
  } catch (error) {
    await signOut(apolloClient);
    throw error;
  }
}

export async function exchangeCodeForToken(
  code: string,
  state: string,
  apolloClient: DollarApollo<any>
) {
  if (!validateState(state)) {
    throw new Error("The state was invalid.");
  }
  const { data } = await apolloClient.mutate({
    mutation: MUTATE_AUTHENTICATE_WITH_GIT_HUB,
    variables: { code: code },
  });
  const authPayload = data.authenticateWithGitHub.authPayload;
  const token = JSON.stringify(authPayload.token);
  return token;
}

export async function getSignInInfo() {
  const { data } = await apolloClient.query({ query: QUERY_SIGN_IN_INFO });
  return _.pick(data, ["isSignedIn", "isAdmin", "gitHubViewer"]);
}
