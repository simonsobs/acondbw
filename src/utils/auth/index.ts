import { Client } from "@urql/vue";
import _ from "lodash";

import { AUTH_TOKEN } from "@/vue-apollo";

import QUERY_SIGN_IN_INFO from "@/graphql/queries/SignInInfo.gql";
import QUERY_IS_SIGNED_IN from "@/graphql/queries/IsSignedIn.gql";

import { exchangeCodeForToken } from "./oauth";

export interface GitHubViewer {
  login: string;
  name?: string;
  avatar_url?: string;
}

export interface SignInInfo {
  gitHubViewer: GitHubViewer | null;
  isSignedIn: boolean;
  isAdmin: boolean;
}

export function restoreFromLocalStorage() {
  try {
    const tokenJson = localStorage.getItem(AUTH_TOKEN);
    const signInInfoJson = localStorage.getItem("sign-in-info");
    const token = tokenJson && (JSON.parse(tokenJson) as string);
    const signInInfo =
      signInInfoJson && (JSON.parse(signInInfoJson) as SignInInfo);

    if (token && signInInfo) {
      return { token, signInInfo };
    }
  } catch {
    localStorage.clear();
  }

  localStorage.removeItem(AUTH_TOKEN);
  localStorage.removeItem("sign-in-info");
  // do not call localStorage.clear() here because it will delete
  // AUTH_STATE when redirected from the auth server

  return {
    token: null as string | null,
    signInInfo: {
      gitHubViewer: null,
      isSignedIn: false,
      isAdmin: false,
    } as SignInInfo,
  };
}

export async function isSignedIn(urqlClient: Client) {
  try {
    const { data } = await urqlClient.query(QUERY_IS_SIGNED_IN, {}).toPromise();
    if (data.isSignedIn) {
      const signInInfo = await getSignInInfo(urqlClient);
      localStorage.setItem("sign-in-info", JSON.stringify(signInInfo));
      return signInInfo;
    }
  } catch {}
  await signOut();
  return false;
}

export async function signOut() {
  // localStorage.removeItem(AUTH_TOKEN);
  localStorage.clear();
}

export async function signIn(
  code: string,
  state: string,
  urqlClient: Client
) {
  try {
    const token = await exchangeCodeForToken(code, state, urqlClient);
    localStorage.setItem(AUTH_TOKEN, token);
    const signInInfo = await getSignInInfo(urqlClient);
    localStorage.setItem("sign-in-info", JSON.stringify(signInInfo));
    return { token, signInInfo };
  } catch (error) {
    await signOut();
    throw error;
  }
}

export async function getSignInInfo(urqlClient: Client) {
  const { data } = await urqlClient.query(QUERY_SIGN_IN_INFO, {}).toPromise();
  return _.pick(data, ["isSignedIn", "isAdmin", "gitHubViewer"]);
}
