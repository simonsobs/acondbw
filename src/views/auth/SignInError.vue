<template>
  <v-container fill-height fluid>
    <v-row align="center" justify="center">
      <v-card flat min-width="250px">
        <v-card-title>Sign In Unsuccessful</v-card-title>
        <v-card-text v-if="error">
          <v-alert type="error" outlined>
            <strong v-if="error.error">{{ error.error }}:</strong>
            <span v-if="error.error_description">
              {{ error.error_description }}</span
            ><v-btn
              v-if="error.error_uri"
              icon
              small
              color="error"
              :href="error.error_uri"
              target="_blank"
              ><v-icon small>mdi-open-in-new</v-icon></v-btn
            >
          </v-alert>
        </v-card-text>
        <v-card-actions>
          <v-btn outlined to="/">Cancel</v-btn>
          <v-spacer></v-spacer>
          <v-btn outlined @click="tryAgain">Try again</v-btn>
        </v-card-actions>
      </v-card>
    </v-row>
  </v-container>
</template>

<script>
export default {
  name: "SignInError",
  computed: {
    error() {
      const auth = this.$store.state.auth;
      if (auth) {
        return auth.lastError;
      } else {
        return null;
      }
    },
  },
  methods: {
    tryAgain() {
      this.$router.push({ name: "SignIn" });
    },
  },
};
</script>
