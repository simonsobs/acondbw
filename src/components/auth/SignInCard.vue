<template>
  <v-progress-circular
    v-if="loading"
    indeterminate
    :size="18"
    :width="3"
    color="secondary"
  ></v-progress-circular>
  <v-card flat v-else>
    <v-card-title>Sign In</v-card-title>
    <v-card-actions>
      <v-btn block outlined @click="signIn">
        <v-icon left>mdi-github</v-icon>Sign In with GitHub
      </v-btn>
    </v-card-actions>
  </v-card>
</template>

<script lang="ts">
import Vue from "vue";
import { v4 as uuidv4 } from "uuid";
import { Location } from "vue-router";
import { mapActions } from "pinia";
import { useAuthStore } from "@/stores/auth";

import {
  redirectToGitHubAuthURL,
  encodeAndStoreState,
  clearState,
  UnencodedState,
} from "@/utils/auth";

export default Vue.extend({
  name: "SignInCard",
  data: () => ({
    loading: false,
  }),
  methods: {
    async signIn() {
      this.loading = true;
      try {
        this.clearAuthError();
        const callbackRoute: Location = { name: "Auth" };
        const scope = ""; // (no scope) https://docs.github.com/en/developers/apps/scopes-for-oauth-apps
        const rawState: UnencodedState = {
          redirect: callbackRoute,
          randomString: uuidv4(),
        };
        const state = encodeAndStoreState(rawState);
        await redirectToGitHubAuthURL(this.$apollo, scope, state);
      } catch (error) {
        clearState();
        this.$router.push({ name: "SignInError" });
        this.loading = false;
      }
    },
    ...mapActions(useAuthStore, { clearAuthError: "clearAuthError" }),
  },
});
</script>
