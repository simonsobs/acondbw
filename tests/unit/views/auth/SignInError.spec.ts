import Vue from "vue";
import VueRouter from "vue-router";
import { PiniaVuePlugin } from "pinia";
import Vuetify from "vuetify";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import { createTestingPinia } from "@pinia/testing";

import SignInError from "@/views/auth/SignInError.vue";

import { useAuthStore } from "@/plugins/pinia/stores/auth";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("SignInError.vue", () => {
  let localVue;
  let vuetify;
  let pinia;
  let authStore;

  function createWrapper() {
    return shallowMount(SignInError, {
      localVue,
      pinia,
      vuetify,
    });
  }

  beforeEach(function () {
    localVue = createLocalVue();
    localVue.use(PiniaVuePlugin);
    pinia = createTestingPinia();
    vuetify = new Vuetify();
    authStore = useAuthStore(pinia);
  });

  it("match snapshot without error", async () => {
    const wrapper = createWrapper();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("match snapshot with error", async () => {
    const wrapper = createWrapper();
    authStore.lastError= {
            error: "access denied",
            error_description: "The user has denied your application access.",
            error_uri:
              "https://docs.github.com/apps/managing-oauth-apps/troubleshooting-authorization-request-errors/#access-denied",
          }
    expect(wrapper.html()).toMatchSnapshot();
  });
});
