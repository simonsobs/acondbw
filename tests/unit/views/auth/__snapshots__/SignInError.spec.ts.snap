// Vitest Snapshot v1

exports[`SignInError.vue > match snapshot with error 1`] = `
"<vuecomponent-stub tag=\\"div\\" fluid=\\"true\\" fill-height=\\"\\">
  <vuecomponent-stub tag=\\"div\\" align=\\"center\\" justify=\\"center\\">
    <vuecomponent-stub loaderheight=\\"4\\" tag=\\"div\\" minwidth=\\"250px\\" flat=\\"true\\">
      <vuecomponent-stub tag=\\"div\\">Sign In Unsuccessful</vuecomponent-stub>
      <!---->
      <vuecomponent-stub tag=\\"div\\">
        <vuecomponent-stub tag=\\"div\\"></vuecomponent-stub>
        <vuecomponent-stub outlined=\\"true\\" tag=\\"button\\" activeclass=\\"\\" to=\\"/\\" type=\\"button\\">Go back</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
  </vuecomponent-stub>
</vuecomponent-stub>"
`;

exports[`SignInError.vue > match snapshot without error 1`] = `
"<vuecomponent-stub tag=\\"div\\" fluid=\\"true\\" fill-height=\\"\\">
  <vuecomponent-stub tag=\\"div\\" align=\\"center\\" justify=\\"center\\">
    <vuecomponent-stub loaderheight=\\"4\\" tag=\\"div\\" minwidth=\\"250px\\" flat=\\"true\\">
      <vuecomponent-stub tag=\\"div\\">Sign In Unsuccessful</vuecomponent-stub>
      <!---->
      <vuecomponent-stub tag=\\"div\\">
        <vuecomponent-stub tag=\\"div\\"></vuecomponent-stub>
        <vuecomponent-stub outlined=\\"true\\" tag=\\"button\\" activeclass=\\"\\" to=\\"/\\" type=\\"button\\">Go back</vuecomponent-stub>
      </vuecomponent-stub>
    </vuecomponent-stub>
  </vuecomponent-stub>
</vuecomponent-stub>"
`;
